### ISSUE TITLE
Programatic Pool Config (like we do for our integration tests but smarter

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#17843]
~~If jdbc/dotCMSPool exists in the App server, use it,~~
~~if db-config.properties exists in the WEB-INF/classes folder, use it~~
~~otherwise we check for secrets and use them,~~
~~otherwise we check for env variables.~~

The order of precedence in which datasource will be initialized is: 
1) `db.properties` in the WEB-INF/classes folder
2) System environments
3) Docker secrets
4) `context.xml` in the `META-INF` folder as always

**Note:** In case you want to use your own implementation, the fully qualified name of the class that implements it has to be set with this property `DATASOURCE_PROVIDER_STRATEGY_CLASS` 

The following properties can be set to define a datasource using options 1, 2 or 3: 
- connection_db_driver --> Default value: `org.postgresql.Driver`
- connection_db_base_url --> Default value: `jdbc:postgresql://localhost/dotcms`
- connection_db_username
- connection_db_password
- connection_db_max_total --> Default value: `60`
- connection_db_max_idle --> Default value: `10`
- connection_db_max_wait --> Default value: `60000`
- connection_db_validation_query --> Default value: `SELECT 1`
- connection_db_leak_detection_threshold --> Default value: `60000`
- connection_db_default_transaction_isolation

**How to set docker secrets:**
We support two ways to set docker secrets:

1. Using a custom file (stored in the docker container) whose absolute path must be set in the `dotmarketing-config.properties` using the property `DOCKER_SECRET_FILE_PATH`. The file must have the following format: 
```
key1=value1
key2=value2
```

2. Using docker commands as explained below. 

**Important Note:** Docker secrets currently are supported only for `02-single-node` implementation because it requires to use `docker stack`, the reason is that `docker stack` doesn't support dependencies among docker services, so the order in which services are deployed is not guaranteed 

**Steps that I followed to create docker secrets and run dotcms with secrets:** 

1) Initialize swarm:
`docker swarm init`
 
2) Create secrets: 
```
printf dotcms | docker secret create connection_db_name -
printf org.postgresql.Driver | docker secret create connection_db_driver -
printf jdbc:postgresql://db:5432/dotcms | docker secret create connection_db_base_url -
printf dotcmsdbuser | docker secret create connection_db_username -
printf password | docker secret create connection_db_password -
printf 60 | docker secret create connection_db_max_total -
printf 10 | docker secret create connection_db_max_idle -
printf 60000 | docker secret create connection_db_max_wait -
printf SELECT 1 | docker secret create connection_db_validation_query -
printf 60000 | docker secret create connection_db_leak_detection_threshold -
```

3) Deploy stack and create containers: 
`docker stack deploy -c docker-compose.yml dotcms_stack`

4) If you want to destroy all the containers created by the stack, use: 
`docker stack rm dotcms_stack`

5) If you need to see the logs in a running container, use: 
`docker logs -f {container_name}`

For further information about how to use Docker Secrets, use the following documentation: 
- https://docs.docker.com/engine/reference/commandline/secret/
- https://docs.docker.com/v17.12/engine/reference/commandline/secret_create/
- https://docs.docker.com/compose/compose-file/#secrets-configuration-reference

As Docker Secrets are supported on Swarm mode only, you might want to learn a bit about it. I used this documentation: 
- https://docs.docker.com/engine/swarm/

**Comment by nollymar** [comment]
Branch in progress: issue-17843-programatic-pool-config

**Comment by nollymar** [comment]
These PRs have been created to support this feature: 
- docker repository (**before merging these PRs, the dotcms image from the `dotcms-src-seed` branch must be generated and published on the repo**) :
https://github.com/dotCMS/docker/pull/19 --> to support Docker Secrets 
https://github.com/dotCMS/docker/pull/18 --> to support external ES

As the `bitnami/postgresql` image does not support Docker Secrets yet, I modified the source of the Postgres image to use the Postgres' official instead. Besides, I created a GitHub issue to ask Bitnami for this implementation https://github.com/bitnami/bitnami-docker-postgresql/issues/195. Once it is implemented, I can roll back the change

- core: 
https://github.com/dotCMS/core/pull/18001

**Comment by dsilvam** [comment]
PR 2: https://github.com/dotCMS/core/pull/18063

**Comment by fabrizzio-dotCMS** [comment]
First I tested that locally that all the possible options for connecting to the database are respected. 

The order by precedence is :
db.properties
Environment variables
And Docker secrets.

If a custom class is provided through the property DATASOURCE_PROVIDER_STRATEGY_CLASS

Such implementation takes place over any of the other methods listed above.
And finally, if none of those methods succeeded one last attempt is made using context.xml 

I was able to verify that order in my local machine.

Here’s the custom class I used.


```
package com.dotcms.util;
import com.dotmarketing.db.DotDataSourceStrategy;
import javax.sql.DataSource;
import org.postgresql.jdbc3.Jdbc3PoolingDataSource;

public class MyDotDataSourceStrategy implements DotDataSourceStrategy {

    @Override
    public DataSource apply() {
        Jdbc3PoolingDataSource source = new Jdbc3PoolingDataSource();
        source.setDataSourceName("A Data Source");
        source.setServerName("localhost");
        source.setDatabaseName("dotcms_sender");
        source.setUser("postgres");
        source.setPassword("admin");
        source.setPortNumber(5433);
        source.setMaxConnections(10);
        return source;
    }
}

```

For the docker secrets part I followed to instructions listed in https://github.com/dotcms/core/issues/17843 under the section

Steps that I followed to create docker secrets and run dotcms with secrets:

But I had to recreate the docker image using a remote branch on which I removed the db.properties. 

In order to do so I had to checkout the docker repository https://github.com/dotCMS/docker

Then I used the following command to build the image locally from the directory 
docker/images/dotcms
Where the Dockerfile is placed.

`docker build --pull --no-cache --build-arg BUILD_FROM=COMMIT --build-arg BUILD_ID=origin/issue-17843-docker-image-testing -t dotcms-test .`

But I also had to rename the file https://github.com/dotCMS/docker/blob/master/images/dotcms/ROOT/srv/templates/dotcms/OVERRIDE/WEB-INF/classes/db.properties

Since that file recreates the db.properties even after it has been deleted from the remote branch. So If you really want to verify that the docker secrets are used you will need to do that too.

The docker image must be run using the command. 

`docker stack deploy -c docker-compose.yml dotcms_stack`

From the directory: /docker/referenceImplementations/02-single_node. Otherwise It won’t work.

So I finally was able to create an image without db.properties and corroborate that the application uses as second option docker secrets since the container doesn’t have any environment variables. But I was able to corroborate the order is followed and that docker secrets are picked up. I even used one of the secrets empty to make sure It wasn’t able to connect. And It didn’t work until I got that fixed.
